### PR TITLE
WIP: Enable kube service account and token controllers

### PR DESCRIPTION
**Pull request by marun**
Depends on https://github.com/kcp-dev/kubernetes/pull/50. Dependency is on the personal branch for that PR until it merges.

- [ ] Check that a service account token can be used to retrieve its containing secret (token is valid, api server can verify)
- [ ] Check that a service account token can not be used to retrieve a secret in an identically named namespace in a different workspace (a token cannot be used across workspaces)

### CODE REVIEW

File: pkg/server/controllers.go [hunk 1]
```diff
@@ -118,6 +124,109 @@ func (s *Server) installKubeNamespaceController(ctx context.Context, config *res
...
+		return nil
+	})
+
+	return nil
+}
+
+func (s *Server) installKubeServiceAccountTokenController(ctx context.Context, config *rest.Config) error {
+	saTokenControllerName := "serviceaccount-token"
+
+	rest.AddUserAgent(config, "tokens-controller")
```
Comment (sttts) [inline pkg/server/controllers.go:161]: 😍 – more of this please

Comment (sttts) [reply pkg/server/controllers.go:161]: don't you have to clone the config first?

File: test/e2e/conformance/token_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	t.Cleanup(cancelFunc)
+
+	server := framework.SharedKcpServer(t)
+	orgClusterName := framework.NewOrganizationFixture(t, server)
+	clusterName := framework.NewWorkspaceFixture(t, server, orgClusterName, "Universal")
+
+	cfg, err := server.DefaultConfig()
+	require.NoError(t, err)
+
+	kubeClient, err := kubernetes.NewClusterForConfig(cfg)
```
Comment (sttts) [inline test/e2e/conformance/token_controller_test.go:49]: kubeClusterClient

File: test/e2e/conformance/token_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	require.NoError(t, err)
+
+	clusterClient := kubeClient.Cluster(clusterName)
```
Comment (sttts) [inline test/e2e/conformance/token_controller_test.go:52]: orgKubeClient

File: test/e2e/conformance/token_controller_test.go (hunk 2 above)
Comment (marun) [reply test/e2e/conformance/token_controller_test.go]: Done

File: test/e2e/conformance/token_controller_test.go (hunk 3 above)
Comment (marun) [reply test/e2e/conformance/token_controller_test.go]: Done

**Comment by marun** [comment]
Is it desirable that kcp automatically generate a service account token signing keypair if not provided? That doesn't appear to be what kube does, but maybe we want to be more user-friendly?

**Comment by sttts** [comment]
> [](https://github.com/marun)[](https://github.com/sttts)[](https://github.com/ncdc)Is it desirable that kcp automatically generate a service account token signing keypair if not provided? That doesn't appear to be what kube does, but maybe we want to be more user-friendly?

All done in https://github.com/kcp-dev/kcp/pull/681. Have integrated your chance into mine.

**Comment by marun** [comment]
Closed in favor of https://github.com/kcp-dev/kcp/pull/681

**Pull request closed without merging**